src/sdk/pynni/nni/compression/torch/gradient_rank_filter_pruners.py

```diff
@@ -0,0 +1,156 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import logging
+import torch
+from .compressor import Pruner
+
+__all__ = ['GradientWeightRankFilterPruner']
+
+logger = logging.getLogger('torch gradient rank filter pruners')
+
+class GradientRankFilterPruner(Pruner):
+    """
+    A structured pruning base class that prunes the filters with the smallest
+    importance criterion in convolution layers (using gradient values)
+    to achieve a preset level of network sparsity.
+    """
+
+    def __init__(self, model, config_list, optimizer, statistics_batch_num=1):
+        """
+        Parameters
+        ----------
+        model : torch.nn.module
+            Model to be pruned
+        config_list : list
+            support key for each list item:
+                - sparsity: percentage of convolutional filters to be pruned.
+        optimizer: torch.optim.Optimizer
+            Optimizer used to train model
+        statistics_batch_num : int
+            Num of batches for calculating contribution
+        """
+
+        super().__init__(model, config_list, optimizer)
+        self.set_wrappers_attribute("if_calculated", False)
+        self.set_wrappers_attribute("contribution", None)
+        self.statistics_batch_num = statistics_batch_num
+        self.iterations = 0
+        self.old_step = self.optimizer.step
+        self.patch_optimizer(self.calc_contributions)
+
+    def calc_contributions(self):
+        raise NotImplementedError('{} calc_contributions is not implemented'.format(self.__class__.__name__))
+
+    def get_mask(self, base_mask, contribution, num_prune):
+        raise NotImplementedError('{} get_mask is not implemented'.format(self.__class__.__name__))
+
+    def calc_mask(self, wrapper, **kwargs):
+        """
+        Calculate the mask of given layer.
+        Filters with the smallest importance criterion which is calculated from the activation are masked.
+
+        Parameters
+        ----------
+        layer : LayerInfo
+            the layer to instrument the compression operation
+        config : dict
+            layer's pruning config
```

Comment:
please update docstring here, i know you copied it here, which means some docstrings somewhere else have not been updated...